cmake/multi_image.cmake

```diff
@@ -0,0 +1,158 @@
+  image_menuconfig(IMAGE_KCONFIG_TARGET IMAGE_COMMAND_TARGET)
+  execute_process(
+    COMMAND ${CMAKE_COMMAND}
+    # Add other toolchain here,
+    -G${CMAKE_GENERATOR}
+    -DBOARD=${IMAGE_BOARD}
+    ${IMAGE_KCONFIG_TARGET}
+    ${IMAGE_COMMAND_TARGET}
+    -DIS_CHILD_IMAGE=True
+    ${sourcedir}
+    ${image_cmake_args}
+    WORKING_DIRECTORY ${${UPNAME}_CMAKE_BINARY_DIR}
+    RESULT_VARIABLE ret
+    )
+  set(${UPNAME}_DOTCONFIG ${${UPNAME}_PROJECT_BINARY_DIR}/.config)
+  set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${${UPNAME}_DOTCONFIG})
+
+  if(NOT ${ret} EQUAL "0")
+    message(FATAL_ERROR "CMake generation for ${name} failed, aborting. Command: ${ret}")
+  endif()
+
+  message("=== child image ${name} end ===")
+
+  # The variables inhereted from PARENT_SCOPE might be overwritten when additional Kconfig is loaded locally.
+  # Therefore, safe guard the variables from parent scope that are needed.
+  import_kconfig(CONFIG_ ${${UPNAME}_PROJECT_BINARY_DIR}/.config)
+
+  # Clear it, just in case it was defined in parent scope
+  set(spm_veneers_lib)
+  if(CONFIG_${UPNAME}_CONTAINS_SECURE_ENTRIES)
+    set(spm_veneers_lib ${${UPNAME}_CMAKE_BINARY_DIR}/${CONFIG_ARM_ENTRY_VENEERS_LIB_NAME})
+  endif()
+
+  define_property(GLOBAL PROPERTY ${name}_KERNEL_NAME
+                 BRIEF_DOCS "Name (KERNEL_NAME) for the SPM image."
+                 FULL_DOCS "To be used to access e.g. this image's hex file."
+                )
+  set_property(GLOBAL PROPERTY ${name}_KERNEL_NAME ${CONFIG_KERNEL_BIN_NAME})
+
+  include(ExternalProject)
+  ExternalProject_Add(${UPNAME}_subimage
+      SOURCE_DIR ${sourcedir}
+      BINARY_DIR ${${UPNAME}_CMAKE_BINARY_DIR}
+      BUILD_BYPRODUCTS "${spm_veneers_lib}"
+                        ${${UPNAME}_PROJECT_BINARY_DIR}/${CONFIG_KERNEL_BIN_NAME}.hex
+      # The CMake configure has been executed above due to cross
+      # dependencies between the multiple CMake configure stages.
+      CONFIGURE_COMMAND ""
+      BUILD_COMMAND ${CMAKE_MAKE_PROGRAM}
+      INSTALL_COMMAND ""
+      BUILD_ALWAYS True
+    )
+
+    # ToDo: Consider to use a dedicated target and populate that target directly
+    # with the files as names and deps.
+    # That would allow to cleanup logic found in partition_manager.cmake
+
+    foreach(kconfig_target
+        menuconfig
+        guiconfig
+        ${EXTRA_KCONFIG_TARGETS}
+        )
+        add_custom_target(${name}_${kconfig_target} ${CMAKE_MAKE_PROGRAM} ${kconfig_target}
+                          WORKING_DIRECTORY ${${UPNAME}_CMAKE_BINARY_DIR}
+                          USES_TERMINAL
+                         )
+    endforeach()
+
+    if (CONFIG_${UPNAME}_CONTAINS_SECURE_ENTRIES AND spm_veneers_lib)
+      # Link the entry veneers library file with the Non-Secure Firmware that needs it.
+      zephyr_link_libraries(${spm_veneers_lib})
+
+      # Add dependency for the benefit of make.
+      add_dependencies(zephyr_interface ${spm_veneers_lib})
```

Comment:
@tejlmand : Do you have steps to reproduce the issue that this line of code fixes?

Reply:
The reason I ask is that I believe that it became redundant with another patch I wrote, and would like to remove it if possible.

Reply:
this line is introduced, because in the original prototype I removed code here:
https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/58ee03b3ab9f9effcfb4d2c5b2bcff616db08f9f/arch/arm/core/cortex_m/tz/CMakeLists.txt#L40
but to minimize upstream `[nrf noup]` then there is no reason to do that, instead we just use the knowledge of:
`CONFIG_${UPNAME}_CONTAINS_SECURE_ENTRIES`.
So this commit https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1508/commits/39007f4caf2e610ca8e745a27e3aa01288745e41fixes this issue:
```
ninja: error: 'libspmsecureentries.a', needed by 'zephyr/zephyr_prebuilt.elf', missing and no known rule to make it
```

Reply:
Note, this dependency is similar in nature to this one: https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/58ee03b3ab9f9effcfb4d2c5b2bcff616db08f9f/arch/arm/core/cortex_m/tz/CMakeLists.txt#L43

Reply:
Let's discuss this in person.